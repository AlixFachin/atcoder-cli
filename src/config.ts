import path from "path";

type Conf = import("conf");

export const defaults = {
	["oj-path"]: undefined,
	["default-contest-dirname-format"]: "{ContestID}",
	["default-task-dirname-format"]: "{tasklabel}"
};

/**
 * 常に単一のConfインスタンスを返す関数
 */
const getConfig: () => Promise<Conf> = (() => {
	// confはsingletonとして扱う
	let conf: Conf | null = null;
	return async () => {
		if (conf !== null) return conf;
		const Conf = (await import("conf")).default;
		return conf = new Conf({defaults});
	}
})();

export default getConfig;

/**
 * コンフィグディレクトリを取得
 */
export async function getConfigDirectory(): Promise<string> {
	return path.resolve((await getConfig()).path, "..");
}
