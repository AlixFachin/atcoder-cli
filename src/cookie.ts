import {name as projectName} from "../package.json";
// yummy!
type CookieConf = import("conf")<Array<string>>;

export interface CookieInterface {
	/**
	 * 現在保持しているcookieを取得
	 */
	get(): Array<string>;
	/**
	 * 現在保持しているcookieを新しいものに置き換える
	 * @param cookies 新しく保持するcookie全体を表す、"key=value"形式の文字列の配列
	 */
	set(cookies: Array<string>): void;
	/**
	 * 現在保持しているcookieを捨てる
	 */
	empty(): void;
	/**
	 * 設定ファイルからcookieを読み込んで保持する
	 */
	loadConfigFile(): Promise<void>;
	/**
	 * 現在保持しているcookieを設定ファイルに書き出す
	 */
	saveConfigFile(): Promise<void>;
}

export interface CookieConstructorInterface {
	/**
	 * 設定ファイルからcookie情報を読み込み済みのインスタンスを生成
	 */
	createLoadedInstance(): Promise<Cookie>
}

/**
 * Cookie管理用クラス
 * cookieはHTTPリクエストヘッダにおける Cookie: a=b; c=d
 * の"a=b", "c=d"の部分を文字列の配列として管理する
 */
export class Cookie implements CookieInterface {
	private cookies: Array<string>;
	private static _cookie_conf: CookieConf | null = null;

	protected static async getCookieConfig(): Promise<CookieConf> {
		if (Cookie._cookie_conf !== null) return Cookie._cookie_conf;
		const Conf = (await import("conf")).default;
		return Cookie._cookie_conf = new Conf({defaults: {cookies: []}, configName: "session", projectName});
	}

	/**
	 * 設定ファイルからcookie情報を読み込み済みのインスタンスを生成
	 */
	static async createLoadedInstance(): Promise<Cookie> {
		const cookie = new Cookie();
		await cookie.loadConfigFile();
		return cookie;
	}

	constructor() {
		this.cookies = [];
	}

	/**
	 * 現在保持しているcookieを取得
	 */
	get(): Array<string> {
		return this.cookies;
	}

	/**
	 * 現在保持しているcookieを新しいものに置き換える
	 * @param cookies 新しく保持するcookie全体を表す、"key=value"形式の文字列の配列
	 */
	set(cookies: Array<string>) {
		this.cookies = cookies;
	}

	/**
	 * Set-Cookieヘッダの配列からこのクラスで管理可能な形式のcookie配列に変換
	 * @param cookies
	 */
	static convertSetCookies2CookieArray(cookies: Array<string>): Array<string> {
		return cookies.map(cookie => /^\s*(?:Set-Cookie:\s*)?(.*)$/i.exec(cookie)![1].trim().split(";")[0])
	}

	/**
	 * 現在保持しているcookieを捨てる
	 */
	empty(): void {
		this.cookies = [];
	}

	/**
	 * 設定ファイルからcookieを読み込んで保持する
	 */
	async loadConfigFile(): Promise<void> {
		this.cookies = (await Cookie.getCookieConfig()).get("cookies");
	}

	/**
	 * 現在保持しているcookieを設定ファイルに書き出す
	 */
	async saveConfigFile(): Promise<void> {
		(await Cookie.getCookieConfig()).set("cookies", this.cookies);
	}
}
