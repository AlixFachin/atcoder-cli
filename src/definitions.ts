export interface Contest {
	id: string;
	title: string;
	url: string;
}

export interface Task {
	id: string;
	label: string;
	title: string;
	url: string;
}

export interface ContestProject {
	contest: Contest;
	tasks: Array<Task>;
}
