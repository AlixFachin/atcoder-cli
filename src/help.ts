export const task_choices = `Supported arguments for --choice:
  inquire  inquire the tasks to add
  all      select all tasks
  none     select no tasks
  rest     select all tasks not added yet
           (without --force option, same with "all")
  next     select the top task that is not added yet`;

export const cmd_env_variables = `Environment variables available in --cmd option:
  $CONTEST_DIR  the path of the contest directory
  $CONTEST_ID   the contest id`;

export const format_strings = `Supported format strings:
  {TaskLabel}     task label (e.g. A, B, C, D, ...)
  {tasklabel}     task label, lower case   
  {TASKLABEL}     task label, upper case
  {TaskID}        task id (shown as task URL path, e.g. arc100_a)
  {TASKID}        task id, upper case
  {TaskTitle}     task title
  {index0}        0-based task index
  {index1}        1-based task index
  {alphabet}      task index as alphabet: a, b, c, ..., z, aa, ab, ...
  {ALPHABET}      task index as alphabet: A, B, C, ..., Z, AA, AB, ...
  {ContestID}     contest id (shown as contest URL path, e.g. arc100)
  {CONTESTID}     contest id, upper case
  {TailNumberOfContestID}  if contest id is arc100, 100 is the tail number. (this format string may be empty, some contests such as abs)
  {ContestTitle}  contest title`;

export const global_config = `Global config options list:
  oj-path                         install path of online-judge-tools (auto detected)
  default-contest-dirname-format  default name of contest directory (created by \`acc new\` command)
  default-task-dirname-format     default name of task directory (created by \`acc new|add\` command)
  default-test-dirname-format     default name of sample cases directory
  default-task-choice             default --choice option for \`acc new|add\` command (see also: \`acc new|add -h\`)
  default-template                default template (see also: \`acc templates -h\`)`;

export const provisioning_templates = `See also:
https://github.com/Tatamo/atcoder-cli#provisioning-templates`;

export const online_judge_tools = `Functions of online-judge-tools linkage:
  downloading sample cases  \`acc new|add\` command
  submit code               \`acc submit\` command`;

export const default_help = `to get detailed information, use \`acc <command> -h\``;
