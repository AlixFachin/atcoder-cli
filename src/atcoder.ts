import {Session} from "./session";
import {ATCODER_BASE_URL, ATCODER_LOGIN_PATH} from "./definitions";
import getConfig from "./config";
import Conf from "conf";
import request, {CookieJar} from "request";

export interface Task {
	id: string,
	label: string,
	title: string,
	url: string
}

export interface Contest {
	id: string,
	title: string,
	url: string
}

export class AtCoder {
	static get base_url(): string {
		return ATCODER_BASE_URL;
	}

	static get login_url(): string {
		return `${AtCoder.base_url}${ATCODER_LOGIN_PATH}`;
	}

	static getContestURL(contest: string) {
		return `${AtCoder.base_url}contests/${contest}`;
	}

	static getTaskURL(contest: string, task: string) {
		return `${AtCoder.getContestURL(contest)}/tasks/${task}`;
	}

	private readonly config: Conf;
	private session: Session;
	// null:未検査 true/false: ログインしているかどうか
	private _login: boolean | null;

	constructor() {
		this.config = getConfig();
		this.session = new Session(this.loadCookiesFromConfig());
		this._login = null;
	}

	/**
	 * ログインしているか調べる
	 * @param force default=false trueならキャッシュを使わずちゃんと調べる
	 */
	async checkSession(force: boolean = false): Promise<boolean> {
		// 以前取得済みならいちいち接続して確かめない
		if (this._login !== null && !force) return this._login;
		return this._login = await this.check();
	}

	/**
	 * アクセスしてログインしている状態かどうかを取得する(結果をキャッシュしない)
	 */
	private async check(): Promise<boolean> {
		// practice contestでログインせず提出ページにアクセスするとコンテストトップに飛ばされることを利用する
		const uri = `${AtCoder.base_url}contests/practice/submit`;
		const response = await this.session.fetch(uri);

		// console.log(response.request.uri.href);
		// リダイレクトされなければログインしている
		return response.request.uri.href === uri;
	}

	/**
	 * ログイン処理します
	 * あまりパスワード文字列を引き回したくないので、この中で標準入力からユーザー名とパスワードを尋ねる
	 */
	async login(): Promise<boolean> {
		if (await this.checkSession()) {
			console.error("you logged-in already");
			return true;
		}
		const csrf_token = await this.getCSRFToken();

		// ユーザーネームとパスワードを入力させる
		const inquirer = await import("inquirer");
		const {username, password} = await inquirer.prompt([{
			type: "input",
			message: "username:",
			name: "username"
		}, {
			type: "password",
			message: "password:",
			name: "password"
		}]) as { username: string, password: string };

		const options = {
			method: "POST",
			formData: {
				username,
				password,
				csrf_token
			}
		};
		const response = await this.session.fetch(AtCoder.login_url, options);
		// ログインページ以外にリダイレクトされていればログイン成功とみなす
		const result = response.request.uri.href !== AtCoder.login_url && response.request.uri.href.indexOf(AtCoder.base_url) === 0;
		if (result) {
			// ログインに成功していた場合はセッション情報を保存する
			this.exportCookiesToConfig(this.session.jar);
		}
		return result;
	}

	/**
	 * ログインページにアクセスしてCSRFトークンを取得
	 */
	private async getCSRFToken(): Promise<string> {
		const {JSDOM} = await import("jsdom");
		const response = await this.session.fetch(AtCoder.login_url);

		const {document} = new JSDOM(response.body).window;
		const input: HTMLInputElement = (document.getElementsByName("csrf_token")[0]) as HTMLInputElement;
		return input.value;
	}

	/**
	 * コンテストIDからコンテストの情報を取得
	 * @param id
	 */
	async contest(id: string): Promise<Contest> {
		const url = AtCoder.getContestURL(id);
		const response = await this.session.fetch(url);
		const {JSDOM} = await import("jsdom");
		const {document} = new JSDOM(response.body).window;
		const regexp = /^(.*) - AtCoder$/;
		const title = regexp.test(document.title) ? regexp.exec(document.title)![1] : document.title;
		return {id, title, url};
	}

	/**
	 * 問題一覧を取得
	 * @param contest
	 */
	async tasks(contest: string): Promise<Array<Task>> {
		const response = await this.session.fetch(`${AtCoder.getContestURL(contest)}/tasks`);

		const {JSDOM} = await import("jsdom");
		const {document} = new JSDOM(response.body).window;
		// very very ad-hoc and not type-safe section
		const tbody = document.querySelector("#main-div .row table>tbody");
		if (tbody === null) return [];
		const tasks: Array<Task> = [];
		for (const tr of tbody.querySelectorAll("tr")) {
			// tr>td>a
			const id: string = tr.children[0].children[0].getAttribute("href")!.split("/").pop()!;
			const label: string = (tr.children[0].children[0] as HTMLAnchorElement).text;
			const title: string = (tr.children[1].children[0] as HTMLAnchorElement).text;
			const url: string = `${AtCoder.base_url}${tr.children[0].children[0].getAttribute("href")!.substring(1)}`;
			tasks.push({id, label, title, url});
		}
		return tasks;
	}

	/**
	 * Configに保存されたCookie情報を読み込む
	 * 引数として与えられたCookieJarの内部状態を書き換えるので、その場合戻り値を再代入する必要はない
	 * @param jar 省略された場合は新しいCookieJarを作って返す
	 */
	private loadCookiesFromConfig(jar?: CookieJar): CookieJar {
		if (jar === undefined) jar = request.jar();
		// configからクッキー情報を取得
		const cookies: string = this.config.get("cookies");
		if (cookies !== undefined) {
			for (const cookie of cookies.split(";")) {
				jar.setCookie(cookie, AtCoder.base_url);
			}
		}
		return jar;
	}

	private exportCookiesToConfig(jar: CookieJar) {
		const cookies = jar.getCookieString(AtCoder.base_url);
		this.config.set("cookies", cookies);
	}
}