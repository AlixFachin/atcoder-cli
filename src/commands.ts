import {AtCoder} from "./atcoder";
import {OnlineJudge} from "./facade/oj";
import {Cookie} from "./cookie";
import * as project from "./project";
import {Contest, Task} from "./definitions";
import getConfig, {defaults} from "./config";
import path from "path";
import {formatTaskDirname, saveProjectJSON} from "./project";

export async function login() {
	const atcoder = new AtCoder();
	console.log(await atcoder.login() ? "OK" : "login failed");
}

export async function logout() {
	// 空のcookieで設定ファイルを上書きする
	await new Cookie().saveConfigFile();
	console.log("login session aborted.");
}

export async function session() {
	const atcoder = new AtCoder();
	console.log("check login status...");
	console.log(await atcoder.checkSession() ? "OK" : "not login");
}

export async function contest(contest_id: string | undefined, options: { id?: boolean }) {
	const f_id = options.id === true;
	const format = ({id, title, url}: Contest) => formatAsShellOutput([[f_id ? SGR(id, 37) : null, SGR(title, 32, 1), url].filter(e => e !== null) as Array<string>]);
	if (contest_id === undefined) {
		// idが与えられていない場合、プロジェクトファイルを探してコンテスト情報を表示
		try {
			const {data: {contest}} = await project.findProjectJSON();
			console.log(format(contest));
		} catch (e) {
			console.error(e.message);
		}
	}
	else {
		try {
			const atcoder = new AtCoder();
			if (!await atcoder.checkSession()) await atcoder.login();
			const contest = await atcoder.contest(contest_id);
			console.log(format(contest));
		} catch {
			console.error(`contest "${contest_id}" not found.`);
		}
	}
}

export async function task(contest_id: string | undefined, task_id: string | undefined, options: { id?: boolean }) {
	const f_id = options.id === true;
	const format = ({id, label, title, url}: Task) => formatAsShellOutput([[f_id ? SGR(id, 37) : null, SGR(label, 32), SGR(title, 32, 1), url].filter(e => e !== null) as Array<string>]);
	if (contest_id === undefined && task_id === undefined) {
		// idが与えられていない場合、プロジェクトファイルを探す
		try {
			const {task} = await project.detectTaskByPath();
			if (task === null) {
				console.error("failed to find the task.");
				return;
			}
			console.log(format(task));
		} catch (e) {
			console.error(e.message);
		}
	}
	else if (contest_id !== undefined && task_id !== undefined) {
		try {
			const atcoder = new AtCoder();
			if (!await atcoder.checkSession()) await atcoder.login();
			const task = await atcoder.task(contest_id, task_id);
			console.log(format(task));
		} catch {
			console.error(`task "${task_id}" of contest "${contest_id}" not found.`);
		}
	}
	else {
		console.error("error: specify both the contest id and the task id.")
	}
}

export async function tasks(contest_id: string | undefined, options: { id?: boolean }) {
	const f_id = options.id === true;
	const format = (tasks: Array<Task>) => formatAsShellOutput(tasks.map(({id, label, title, url}) => [f_id ? SGR(id, 37) : null, SGR(label, 32), SGR(title, 32, 1), url].filter(e => e !== null) as Array<string>));
	if (contest_id === undefined) {
		// idが与えられていない場合、プロジェクトファイルを探す
		try {
			const {data: {tasks}} = await project.findProjectJSON();
			console.log(format(tasks));
		} catch (e) {
			console.error(e.message);
		}
	}
	else {
		try {
			const atcoder = new AtCoder();
			if (!await atcoder.checkSession()) await atcoder.login();
			const tasks = await atcoder.tasks(contest_id);
			console.log(format(tasks));
		} catch {
			console.error(`contest "${contest_id}" not found.`);
		}
	}
}

export async function url(contest_id: string | undefined, task_id: string | undefined, options: { check?: boolean }) {
	const f_check = options.check === true;
	if (contest_id !== undefined && task_id !== undefined) {
		if (f_check) {
			const atcoder = new AtCoder();
			try {
				const tasks = await atcoder.tasks(contest_id);
				// Task一覧から一致するURLを探す
				for (const {url} of tasks) {
					if (url === AtCoder.getTaskURL(contest_id, task_id)) {
						console.log(url);
						return;
					}
				}
				// なかった
				console.error(`task "${task_id}" not found.`);
			} catch {
				console.error(`contest "${contest_id}" not found.`);
			}
		}
		else {
			// URLの妥当性をチェックしない
			console.log(AtCoder.getTaskURL(contest_id, task_id));
		}
	}
	else if (contest_id !== undefined && task_id === undefined) {
		if (f_check) {
			const atcoder = new AtCoder();
			try {
				// コンテストページの存在確認
				const {url} = await atcoder.contest(contest_id);
				console.log(url);
			} catch {
				console.error(`contest "${contest_id}" not found.`);
			}
		}
		else {
			// URLの妥当性をチェックしない
			console.log(AtCoder.getContestURL(contest_id));
		}
	}
	else {
		console.log(AtCoder.base_url);
	}
}

export async function submit(filename: string, options: { task?: string, contest?: string }) {
	let contest_id = options.contest;
	let task_id = options.task;
	if (contest_id === undefined || task_id === undefined) {
		// コンテストまたはタスクが未指定の場合、カレントディレクトリのパスから提出先を調べる
		const {contest, task} = await project.detectTaskByPath();
		if (contest_id === undefined && contest !== null) contest_id = contest.id;
		if (task_id === undefined && task !== null) task_id = task.id;

		// 結局特定できなかった
		if (contest_id === undefined || task_id === undefined) {
			console.error(`cannot find the ${task_id === undefined ? "task" : "contest"} to submit.`);
			console.error(`add ${task_id === undefined ? "-t" : "-c"} flag to specify it.`);
			return;
		}
	}

	if (!await OnlineJudge.checkAvailable()) {
		console.error("online-judge-tools is not available.");
		return;
	}

	// URLの妥当性をチェック
	const url: string | null = (await new AtCoder().task(contest_id, task_id).catch(() => ({url: null}))).url;
	if (url === null) {
		console.error(`Task ${AtCoder.getTaskURL(contest_id, task_id)} not found.`);
		return;
	}
	console.log(`submit to: ${url}`);
	// 提出
	await OnlineJudge.call(["s", url, filename]);
}

export async function checkOJAvailable() {
	const available = await OnlineJudge.checkAvailable();
	const path = await OnlineJudge.getPath();
	console.log(`online-judge-tools is ${available ? "" : "not "}available. ${available ? "found at:" : ""}`);
	if (available) {
		console.log(path);
	}
}

export async function configDir() {
	const conf = await getConfig();
	console.log(path.resolve(conf.path, ".."));
}

export async function config(key: string | undefined, value: string | undefined, options: { D?: boolean }) {
	if (options.D) {
		await deleteGlobalConfig(key);
	}
	else if (key !== undefined && value !== undefined) {
		await setGlobalConfig(key, value);
	}
	else {
		await getGlobalConfig(key);
	}
}

async function getGlobalConfig(key?: string) {
	const conf = await getConfig();
	if (key === undefined) {
		for (const key of Object.keys(defaults)) {
			console.log(undef2empty`${key}: ${conf.get(key)}`)
		}
		return;
	}
	if (!(key in defaults)) {
		console.error(`invalid option "${key}".`);
		return;
	}
	console.log(undef2empty`${conf.get(key)}`);
}

async function setGlobalConfig(key: string, value: string) {
	const conf = await getConfig();
	if (!(key in defaults)) {
		console.error(`invalid option "${key}".`);
		return;
	}
	conf.set(key, value);
	console.log(undef2empty`${key} = ${conf.get(key)}`);
}

async function deleteGlobalConfig(key?: string) {
	const conf = await getConfig();
	if (key !== undefined) {
		if (!(key in defaults)) {
			console.error(`invalid option "${key}".`);
			return;
		}
		conf.delete(key);
		console.log(`option "${key}" is set back to default.`);
	}
	else {
		console.error("option key is not specified.");
	}
}

/**
 * テンプレート文字列に挿入された式がundefinedであった場合に"undefined"のかわりに空文字列に変換する
 * @param strings
 * @param values
 */
function undef2empty(strings: TemplateStringsArray, ...values: Array<any>): string {
	values = values.map(value => value !== undefined ? value : "");
	return String.raw(strings, ...values);
}

export async function setup(contest_id: string, options: { choice: "inquire" | "all" | "none" | "rest" | "next", force?: boolean, contestDirnameFormat?: string, taskDirnameFormat?: string }) {
	try {
		const {contest} = await project.init(contest_id, options);
		console.log(`create project of ${contest.title}`);
		await add(options);
	} catch (e) {
		console.error(e.message);
	}
}

export async function add(options: { choice: "inquire" | "all" | "none" | "rest" | "next", force?: boolean, taskDirnameFormat?: string }) {
	try {
		const {path, data} = await project.findProjectJSON();
		const {contest, tasks} = data;
		const choices = await selectTasks(tasks, options.choice, options.force);
		for (const {index, task} of choices) {
			// forceオプションが設定されていない場合、既にディレクトリが存在する問題はスキップする
			if (options.force !== true && task.directory !== undefined) continue;
			const format = options.taskDirnameFormat !== undefined ? options.taskDirnameFormat : (await getConfig()).get("default-task-dirname-format");
			const dirname = formatTaskDirname(format, task, index, contest);
			// 新しいTaskが返ってくるので、もともとの配列の要素を更新する
			tasks[index] = await project.installTask(task, dirname, path);
		}
		// 更新されたContestProjectをファイルに書き出し
		await saveProjectJSON(Object.assign(data, {tasks}));
	} catch (e) {
		console.error(e.message);
	}
}

export async function selectTasks(tasks: Array<Task>, choice: "inquire" | "all" | "none" | "rest" | "next", force: boolean = false): Promise<Array<{ index: number, task: Task }>> {
	switch (choice) {
		case "inquire":
			return await inquireTasks(tasks, force);
		case "all":
			return tasks.map((task, index) => ({index, task}));
		case "none":
			return [];
		case "rest":
			return tasks.filter(task => task.directory === undefined).map((task, index) => ({index, task}));
		case "next":
			const next = getNextTask2Install(tasks);
			return next !== null ? [next] : [];
		default:
			throw new Error(`mode "${choice}" is not defined"`);
	}
}

export async function inquireTasks(tasks: Array<Task>, force: boolean = false): Promise<Array<{ index: number, task: Task }>> {
	const inquirer = await import("inquirer");
	// まだディレクトリが作成されていない問題を一つだけ選択状態にしておく
	const next = getNextTask2Install(tasks);
	if (!force && next === null) throw new Error("all tasks are already installed. use --force option to override them.");
	return (await inquirer.prompt([{
		type: "checkbox",
		message: "select tasks",
		name: "tasks",
		choices: tasks.map((task, index) => ({
			name: `${task.label} ${task.title}`,
			value: {index, task},
			disabled: !force && task.directory !== undefined ? () => "already installed" : () => false,
			checked: next !== null && index === next.index
		}))
	}]) as { tasks: Array<{ index: number, task: Task }> }).tasks;
}

/**
 * まだディレクトリが作成されていない問題のうち、最も上のものを得る
 * すべての問題のディレクトリが作成済みの場合はnullを返す
 * @param tasks
 */
function getNextTask2Install(tasks: Array<Task>): { index: number, task: Task } | null {
	for (let i = 0; i < tasks.length; i++) {
		if (tasks[i].directory === undefined) {
			return {index: i, task: tasks[i]};
		}
	}
	return null;
}

/**
 * 文字列の2次元配列を受け取り、行ごとにいい感じのスペースで結合して返す
 * @param data
 */
function formatAsShellOutput(data: Array<Array<string>>): string {
	const padding = "  ";
	const max_lengths = data.map(arr => arr.map(str => str.length)).reduce((a, b) => a.map((v, i) => Math.max(v, b[i])));
	return data.map(line => line.reduceRight((p, c, i) => c.padEnd(max_lengths[i]) + padding + p)).join("\n");
}

/**
 * プロセスがTTYで実行されているならSGRでスタイルを適用する
 * @param str 適用する文字列
 * @param codes コード番号(可変長)
 */
function SGR(str: string, ...codes: Array<number>) {
	if (!process.stdout.isTTY) return str;
	return `${codes.map(code => `\x1b[${code}m`).join("")}${str}\x1b[0m`;
}
