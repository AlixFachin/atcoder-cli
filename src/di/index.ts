import {Design, Injector} from "typesafe-di";
import { Session } from "../session";
import { AtCoder } from "../atcoder";
import { Cookie } from "../cookie";

interface HasCookieConstructor {
    CookieConstructor: typeof Cookie;
}
interface HasSession {
    session: Session;
}

export const CookieDesign = Design.bind('CookieConstructor', ()=> Cookie)

export const SessionDesign = Design.bind('session', async (injector: Injector<HasCookieConstructor>) => 
    new Session(await injector.CookieConstructor)
)

export const AtCoderDesign = Design.bind('atcoder', async (injector: Injector<HasSession>) =>
    new AtCoder(await injector.session)
)

export const productionAtCoderDesign = AtCoderDesign.merge(CookieDesign).merge(SessionDesign)

/**
 * Generate an AtCoder instance for production.
 */
export const getAtCoder = async () => {
    const {
        container: { atcoder }
    } = await productionAtCoderDesign.resolve({});
    return atcoder
};


