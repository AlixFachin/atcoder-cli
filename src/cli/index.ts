import commander from "commander";
// import {version} from "../../package.json";
// work-around to avoid TS5505 error bug in TypeScript 3.1 https://github.com/Microsoft/TypeScript/issues/24715
const {version} = require("../../package.json");
import * as commands from "../commands";

commander
	.version(version, "-v, --version");

commander
	.command("new <contest-id>")
	.alias("setup")
	.action(commands.setup)
	.description("create new contest project directory");

commander
	.command("login")
	.action(commands.login)
	.description("login to AtCoder");

commander
	.command("logout")
	.action(commands.logout)
	.description("delete login session information");

commander
	.command("session")
	.action(commands.session)
	.description("check login or not");

commander
	.command("contest [contest-id]")
	.action(commands.contest)
	.option("-i, --id", "show contest id")
	.description("get contest information");

commander
	.command("tasks [contest-id]")
	.action(commands.tasks)
	.option("-i, --id", "show task id")
	.description("get tasks");

commander
	.command("url [contest] [task]")
	.option("-c, --check", "check the specified contest and/or task id is valid")
	.action(commands.url)
	.description("get contest or task URL");

commander
	.command("oj")
	.action(commands.oj)
	.description("call online-judge");

commander.parse(process.argv);
