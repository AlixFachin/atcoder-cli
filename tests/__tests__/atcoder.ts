jest.useFakeTimers();
import {username, password} from "./auth.json";
import {AtCoder} from "../../src/atcoder";
import {disableCookieFileIO, mockLogin} from "../utils";
import { productionAtCoderDesign } from "../../src/di/index.js";

// ログイン情報が実際にコンフィグファイルに書き込まれないようにする
disableCookieFileIO();

// テスト用のAtCoderインスタンスを生成（現時点ではまだproduction用と同じ）
const getTestAtCoder = async () => {
	const { container: { atcoder }} = await productionAtCoderDesign.resolve({});
	return atcoder;
}

/*
 このテストが失敗する場合、まず以下の点を確認してください
 - __tests__/auth.json が存在し、正しいユーザー名とパスワードが記述されていること
   すべてのテストを開始する前に、__tests__/auth.example.jsonをコピーして__tests__/auth.json を作成し、
   有効なAtCoderアカウントのユーザー名とパスワードを記述してください
   __tests__/auth.jsonはgit管理に含めないように注意してください
 */
test("AtCoder Login", async () => {
	const atcoder = await getTestAtCoder();
	expect(await atcoder.checkSession()).toBe(false);
	expect(await mockLogin(atcoder, {username, password})).toBe(true);
	expect(await atcoder.checkSession(true)).toBe(true);
});

describe("AtCoder get information", () => {
	// 使用前にbeforeAllで代入される
	let atcoder!: AtCoder;
	beforeAll(async () => {
		if (!atcoder) {
			atcoder = await getTestAtCoder();
		}
		await mockLogin(atcoder, {username, password});
	});
	describe("contest and tasks", ()=> {
		const contests = ["abc101", "arc101"];
		describe("contest", () => {
			test.each(contests)("%s", async (contest_id) => {
				expect(await atcoder.contest(contest_id)).toMatchSnapshot();
			});
			test("invalid contest id", async () => {
				await expect(atcoder.contest("abc0xx")).rejects.toThrow();
			});
		});
		describe("tasks", () => {
			test.each(contests)("%s", async (contest_id) => {
				expect(await atcoder.tasks(contest_id)).toMatchSnapshot();
			});
			test("invalid contest id", async () => {
				await expect(atcoder.tasks("abc0xx")).rejects.toThrow();
			});
		});
		const tasks = [["abc101", "abc101_a"], ["abc101", "abc101_b"], ["arc101", "arc101_a"]];
		describe("task", () => {
			test.each(tasks)("%s %s", async (contest_id, task_id) => {
				expect(await atcoder.task(contest_id, task_id)).toMatchSnapshot();
			});
			test("invalid contest id", async () => {
				await expect(atcoder.task("abc0xx", "abc0xx_z")).rejects.toThrow();
			});
			test("invalid task id", async () => {
				await expect(atcoder.task("abc101", "abc102_a")).rejects.toThrow();
			});
		});
	});
});
