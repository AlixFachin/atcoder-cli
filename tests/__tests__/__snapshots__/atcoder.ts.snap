// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AtCoder get information contest and tasks contest aic987 1`] = `
Object {
  "id": "aic987",
  "title": "AtCoder Imaginary Contest 987",
  "url": "https://atcoder.jp/contests/aic987",
}
`;

exports[`AtCoder get information contest and tasks task aic987 aic987_a 1`] = `
Object {
  "id": "aic987_a",
  "label": "A",
  "title": "This is Problem A",
  "url": "https://atcoder.jp/contests/aic987/tasks/aic987_a",
}
`;

exports[`AtCoder get information contest and tasks task aic987 aic987_b 1`] = `
Object {
  "id": "aic987_b",
  "label": "B",
  "title": "Next Problem",
  "url": "https://atcoder.jp/contests/aic987/tasks/aic987_b",
}
`;

exports[`AtCoder get information contest and tasks task aic987 aic987_c 1`] = `
Object {
  "id": "aic987_c",
  "label": "C",
  "title": "The Test Cases",
  "url": "https://atcoder.jp/contests/aic987/tasks/aic987_c",
}
`;

exports[`AtCoder get information contest and tasks tasks aic987 1`] = `
Array [
  Object {
    "id": "aic987_a",
    "label": "A",
    "title": "This is Problem A",
    "url": "https://atcoder.jp/contests/aic987/tasks/aic987_a",
  },
  Object {
    "id": "aic987_b",
    "label": "B",
    "title": "Next Problem",
    "url": "https://atcoder.jp/contests/aic987/tasks/aic987_b",
  },
  Object {
    "id": "aic987_c",
    "label": "C",
    "title": "The Test Cases",
    "url": "https://atcoder.jp/contests/aic987/tasks/aic987_c",
  },
  Object {
    "id": "aic987_d",
    "label": "D",
    "title": "Imaginary Problem",
    "url": "https://atcoder.jp/contests/aic987/tasks/aic987_d",
  },
]
`;
